### ISSUE TITLE
Is not using .js files on servers quite dangerous and prone to cause vulnerabilities?

### ISSUE THREAD

**Issue by rugk** [small-tech/jsdb#11]
This all sounds nice to store data in `.js` files directory, however, is not this quite dangerous when you think about it?

## The problem

Because as soon as you use it for some sensitive data of some kind – which almost always happens when you use a database, even [aggregations of what seems to be quite public data](https://www.troyhunt.com/8-million-github-profiles-were-leaked-from-geekedins-mongodb-heres-how-to-see-yours/) can be sensitive – it gets somewhat critical. (And BTW; you talk about privacy in your Readme, so clearly you consider the fact that there could be private data in the DBs.)
The technical issue of course is easy: By default, when `.js`  files are placed on a webserver, anyone can download them, as `.js` files are of course intended to be used like databases.

This is not at all theoretical. After all data leaks [are very common](https://haveibeenpwned.com/PwnedWebsites) and simply downloading files from random webservers under well-known locations [is somewhat shockingly incredible common, so that even tools exist to automate that](https://github.com/hannob/snallygaster). See also [the list of breaches and linked news articles here which have all that in common](https://github.com/hannob/snallygaster/blob/main/TESTS.md). And people [forget `.git` directories](https://github.com/internetwache/GitTools) or saved `nano`/`vim` backup copies of files… so it is easy to forget. Or database dumps… etc. [See the also linked talk e.g.](https://www.youtube.com/watch?v=Bppr9rbmwz4).
I could go on here, talk about [a vulnerability/problem in a project I am affiliated with](https://privatebin.info/reports/vulnerability-2017-09-29.html), where we ended up using a `config.ini.php`  instead of `config.ini` file, because that is guaranteed to be parsed by most web servers by default and thus won't expose sensitive data. (That's also why most public CMSes based on PHP e.g. do it like this…) 

## Solution?

It may be obvious to you that of course you should not use the project like this/keep the database files private, but _stating the obvious_ is always a good idea when it comes to security.
As such, _at least_ document that you (obviously, as said, but well… keeping `.ini` files private is also obvious, yet it happens that these are exposed due to accidental misconfiguration etc.…) should somehow place the files into a directory not-accessible by any webserver. Maybe you can also do some further hardening with user permissions, SeLinux or webserver onfiguration…
E.g. what may an – arguably ugly – workaround is if you save your files as `database.js.php`, which makes the files being non-accessible if they are on a PHP-compatible webserver. Otherwise, a different file ending such as `.jsdb` or another not-served by webservers file extension may be used – though of course the protection effect depends on how webservers treat such a file extension _by default_. (After all, remember people are lazy… manually adjusting webserver configs is likely not being done. :wink:)

Also, of course, it should be remembered people may take this into care when directly working with this here, respectively, if they know they work with that here. However, when used in third-party projects like your Site.js that information/risk can soon be forgotten about.

**Comment by aral** [comment]
@rugk Good point on documenting that the database files should not be kept in web directories. However, in this regard, it is no more or less secure than any other database that keeps its data in files. So the same concern applies to SQLite, for example. There’s nothing inherently more dangerous to have the data in JS files apart from the fact that you could easily query the data when it loads in a browser vs, say, having to launch SQLite to do so after you’ve downloaded the database. Either way, if you serve a database file from your web server, the data will be compromised.

PS. I’m about to move this repository to Codeberg. I’ll make sure the documentation issue is tracked there.

**Comment by aral** [comment]
Just closing off issues here.

[Kitten](https://kitten.small-web.org) is a good reference for how to use JSDB securely in a web context.

(As mentioned above, do not keep JSDB databases – or any other database files of other databases, for that matter – inside your web directories. This is an issue that applies equally to all files that contain sensitive information, regardless of format.)

**Comment by rugk** [comment]
Thanks for the ping.

>PS. I’m about to move this repository to Codeberg. I’ll make sure the documentation issue is tracked there.

Ah I see I created https://codeberg.org/small-tech/jsdb/issues/35

> it is no more or less secure than any other database that keeps its data in files
> This is an issue that applies equally to all files that contain sensitive information, regardless of format. 

Or indeed obviously is, and I elaborated in the original issue that this is a very common issue, [as breaches are in general](https://haveibeenpwned.com/PwnedWebsites). The thing is: As a maintainer of the software you can try to minimize the impact:
* make secure defaults that cannot be downloaded e.g. PHP software often stored their ini configuration data in a PHP file that is interpreted by the host, so even in a misconfiguration, which, again, happens, it is not served to the clients. Scans alwqays show, people [will misconfogire it](https://privatebin.info/reports/vulnerability-2017-09-29.html).
* that's why it's also important to document that. As said state the obvious!
* things like [default passwords in software](https://owasp.org/www-project-top-10-infrastructure-security-risks/docs/2023/INT07_2023-Insecure_Passwords_and_Default_Credentials) e.g. are, for that reason, [still a big problem](https://dev.to/guardingpearsoftware/why-default-passwords-are-still-a-massive-problem-in-2025-3gc4)
* Also, the comparison with other db systems is not quite fair, as there are differences as: other db's place their files when installed, by default, in some other system directory and not in the root of itself. And good dbms systems auto-generated credentials upon first start, so no default passwords are there should that service end up being exposed to the network. 
  * as said not all do that and redis is an infamous example, which then explains how really way to o many redis instances are open on the Internet:
  > 
    There are 39,405 unauthenticated Redis services out of 350,675 total Redis services on the public internet.
  >  Almost 50% of unauthenticated Redis services on the internet show signs of an attempted compromise.
  [As the blog post also argues,](https://censys.com/blog/databases-exposed-redis) I'd argue this is also due to the insecure defaults being used there.
* what you can also do is ensure these secure default locations in code (at least by default), respectively log an error when insecure ones are used.
